File: go/marsrover/mission.go
package marsrover

import (
  "strings"
  "strconv"
  "fmt"
)

// Range represents exploration mission range
type Range struct {
  X int
  Y int
}

// Mission represents exploration mission
type Mission struct {
  Plan   []string
  Result []string
  Range  Range
  rover  Rover
  step   int
}

// Rover represents rover placement
type Rover struct {
  X         int
  Y         int
  Direction rune
}

func (rover *Rover) String() string {
  return fmt.Sprint(rover.X, rover.Y, " ", string(rover.Direction))
}

// Explore does exploration
func (mission *Mission) Explore() {
  if mission.hasPlan() {
    mission.start()
    mission.readRange()
    for mission.continues() {
      mission.landRover()
      if mission.continues() {
        mission.commandRover()
      }
      mission.reportRover()
    }
  }
}

func (mission *Mission) hasPlan() bool {
  return len(mission.Plan) > 0
}

func (mission *Mission) start() {
  mission.step = 0
}

func (mission *Mission) continues() bool {
  return mission.step < len(mission.Plan)
}

func (mission *Mission) readRange() {
  items := strings.SplitN(mission.Plan[0], " ", 2)
  x,_ := strconv.Atoi(items[0])
  y,_ := strconv.Atoi(items[1])
  mission.Range = Range { X: x, Y: y }
  mission.step++
}

func (mission *Mission) landRover() {
  record := mission.Plan[mission.step]
  items := strings.SplitN(record, " ", 3)
  x,_ := strconv.Atoi(items[0])
  y,_ := strconv.Atoi(items[1])
  dir := []rune(items[2])[0]
  mission.rover = Rover { X: x, Y: y, Direction: dir}
  mission.step++
}

func (mission *Mission) commandRover() {
  instructions := mission.Plan[mission.step]
  for _, cmd := range instructions {
    switch cmd {
    case 'L': { mission.spinRoverLeft() }
    case 'R': { mission.spinRoverRight() }
    case 'M': { mission.moveRover() }
    }
  }
  mission.step++
}

func (mission *Mission) spinRoverLeft() {
  switch mission.rover.Direction {
  case 'N': { mission.rover.Direction = 'W' }
  case 'W': { mission.rover.Direction = 'S' }
  case 'S': { mission.rover.Direction = 'E' }
  case 'E': { mission.rover.Direction = 'N' }
  }
}

func (mission *Mission) spinRoverRight() {
  switch mission.rover.Direction {
  case 'N': { mission.rover.Direction = 'E' }
  case 'E': { mission.rover.Direction = 'S' }
  case 'S': { mission.rover.Direction = 'W' }
  case 'W': { mission.rover.Direction = 'N' }
  }
}

func (mission *Mission) moveRover() {
  switch mission.rover.Direction {
  case 'N': { mission.rover.Y++ }
  case 'S': { mission.rover.Y-- }
  case 'E': { mission.rover.X++ }
  case 'W': { mission.rover.X-- }
  }
}

func (mission *Mission) reportRover() {
  mission.Result = append(mission.Result, mission.rover.String())
}

File: csharp/marsrover/Mission.cs
using System.Collections.Generic;

namespace Marsrover
{
  public class Mission
  {
    private List<string> result = new List<string>();

    public string[] Plan { private get; set; }

    public RangeType Range { get; private set; }

    public IEnumerable<string> Result { get => result; }

    private Rover rover;

    public struct RangeType
    {
      public int X { get; set; }
      public int Y { get; set; }
    }

    public struct Rover
    {
      public int X { get; set; }

      public int Y { get; set; }

      public char Direction { get; set; }
    }

    public void Explore()
    {
      if (Plan != null && Plan.Length > 0)
      {
        ReadRange();
        if (Plan.Length > 1)
        {
          LandRover();
          if (Plan.Length > 2)
          {
            CommandRover();
          }
          ReportRover();
        }
      }
    }

    private void ReadRange()
    {
      var fields = Plan[0].Split(" ", 2);
      Range = new RangeType
      {
        X = int.Parse(fields[0]),
        Y = int.Parse(fields[1]),
      };
    }

    private void LandRover()
    {
      var fields = Plan[1].Split(" ", 3);
      rover = new Rover {
        X = int.Parse(fields[0]),
        Y = int.Parse(fields[1]),
        Direction = fields[2][0],
      };
    }

    private void CommandRover()
    {
      foreach (var command in Plan[2])
      {
        CommandRover(command);
      }
    }

    private void CommandRover(char command)
    {
      switch (command)
      {
        case 'R': SpinRoverRight(); break;
        case 'L': SpinRoverLeft(); break;
        case 'M': MoveRover(); break;
      }
    }

    private void MoveRover()
    {
      switch (rover.Direction)
      {
        case 'E': rover.X++; break;
        case 'W': rover.X--; break;
        case 'N': rover.Y++; break;
        case 'S': rover.Y--; break;
      }
    }

    private void SpinRoverRight()
    {
      switch (rover.Direction)
      {
        case 'N': rover.Direction = 'E'; break;
        case 'E': rover.Direction = 'S'; break;
        case 'S': rover.Direction = 'W'; break;
        case 'W': rover.Direction = 'N'; break;
      }
    }

    private void SpinRoverLeft()
    {
      switch (rover.Direction)
      {
        case 'N': rover.Direction = 'W'; break;
        case 'W': rover.Direction = 'S'; break;
        case 'S': rover.Direction = 'E'; break;
        case 'E': rover.Direction = 'N'; break;
      }
    }

    private void ReportRover()
    {
      result.Add($"{ rover.X } { rover.Y } { rover.Direction }");
    }
  }
}
